### PR TITLE
Issue 215 update alpine

### PR DESCRIPTION
**Pull request by nigelgbanks**
~This isn't working yet, hoping to have time soonish, but just wanted to push this out on the radar so that work isn't duplicated.~

Ready for review.

### CODE REVIEW

**Comment by seth-shaw-asu** [comment]
I think this branch needs to be updated. The first commit has already been integrated into main.

Today I was able to get a Drupal container running PHP 8.1 by starting off with the second commit and then making a few additional changes, mostly to fpm paths ('.txt' extension added so GitHub will upload the file): [php81-20220912.patch.txt](https://github.com/Islandora-Devops/isle-buildkit/files/9559864/php81-20220912.patch.txt)

**Comment by seth-shaw-asu** [comment]
I did a bit more PHP 8.1 clean-up. See my working branch: https://github.com/asulibraries/isle-buildkit/tree/asu.

**Comment by nigelgbanks** [comment]
When this is finally testable, note the following [issue](https://github.com/Islandora/islandora-starter-site/issues/19) where in fits doesn't get generated on media creation, one can still generate it by manually applying the action and the bug resides out of this repository's hands, so you can ignore that bit.

**Comment by nigelgbanks** [comment]
In the included test site, there is no search displays setup, that will come at a later date when the following [issue](https://github.com/Islandora/islandora-starter-site/issues/23) is resolved, in the meanwhile you can visit the [solr.islandora.dev] to confirm documents are actually indexing.

**Comment by nigelgbanks** [comment]
Automating the installation of Matomo via the console using [ExtraTools](https://github.com/digitalist-se/extratools) is just too unstable and flaky that I don't we can rely on it.

**Comment by seth-shaw-asu** [comment]
> Automating the installation of Matomo via the console using [ExtraTools](https://github.com/digitalist-se/extratools) is just too unstable and flaky that I don't we can rely on it.

Oh, bother, I ended up not using ExtraTools anyway but forgot to take it out! 🤦‍♂️

File: composer/Dockerfile [hunk 1]
```diff
@@ -1,11 +1,11 @@
-# syntax=docker/dockerfile:1.2.1
+# syntax=docker/dockerfile:1.4.3
 ARG repository=local
 ARG tag=latest
-ARG alpine=3.15.0
-FROM --platform=$BUILDPLATFORM ${repository}/download:${tag} AS download 
+ARG alpine=3.16.2
```
Comment (misilot) [inline composer/Dockerfile:4]: Instead of specifying the patch version would it make sense to just target the minor version?

Comment (nigelgbanks) [reply composer/Dockerfile:4]: In this case, no. As that could produce different images depending on the time in which it is executed, even for the same commit to this repository. Ideally, a given commit in `isle-buildkit` should always produce the same image if built locally, remotely, now or in three months in the future. Our goal being stable reproducible builds.

**Comment by nigelgbanks** [comment]
Needed to re-write the plugin to better support the automated test / improving caching and speed up build times.

https://github.com/Islandora-Devops/isle-gradle-docker-plugin/compare/0.11...1.0.1

I can now do a complete clean build (no-cache) ~8 mins on my MacBook, whereas before it took around 35 minutes. With caching, about ~4 mins on my MacBook.

**Comment by nigelgbanks** [comment]
This is why the tests are failing: https://github.com/just-containers/s6-overlay/issues/503, in the meanwhile. I'll raise a ticket to patch when the new release of s6-overlay comes out, and simply ignore when services do not have enough time to react to the SIGTERM. This by and large only effects the JAVA services (ActiveMQ, Fcrepo, etc) which is fortunate in that they are typically the only service running in the container (so not affected by the bug), except when we are also running tests, as the tests are additional services.

**Comment by nigelgbanks** [comment]
I am not sure why I'm getting cache misses with GitHub actions that I don't see locally...

**Comment by DonRichards** [comment]
We had to explore this recently. I don't know if this helps, but this is what we're doing to get to PHP 8.1 https://github.com/jhu-idc/idc-isle-buildkit/pull/112

**Comment by nigelgbanks** [comment]
> We had to explore this recently. I don't know if this helps, but this is what we're doing to get to PHP 8.1 [jhu-idc#112](https://github.com/jhu-idc/idc-isle-buildkit/pull/112)

Oh I've had PHP 8.1 working for months, the things I have had issues with are more related to working around bugs in Fedora (locked up and didn't respond to SIGTERM singles causing the JVM to hang - though I've resolved this) and buildkit (Issues relate to caching and concurrency bugs -  I'm still working on this)

**Comment by nigelgbanks** [comment]
I had a meeting with @rosiel to go over these changes, and she'll continue the code review later. While 100% of everything is not polished, things are working and building, and it's better to get it in and make incremental changes going forward than keeping this pull request up indefinitely.

**Comment by DonRichards** [comment]
@nigelgbanks Do we know when this will be ready?

**Comment by nigelgbanks** [comment]
@DonRichards It's ready for review now, I'll be continuing my work on other branches adding more docs, etc.

**Comment by jefferya** [comment]
A note about my experience testing.

In context of a shared storage environment, the mariadb container fails to complete `mysql-setup` (newly created volume). An example of the failure is below. However, If I set the docker-compose.yml env var `S6_CMD_WAIT_FOR_SERVICES_MAXTIME` for mariadb then the mysql-setup initialization succeeds. Also Initialization succeeds when used on a different system with local SSD storage. 

Docker version 23.0.1, build a5ee5b1
Docker Compose version v2.16.0
Image Tag: islandora/mariadb:issue-215-update-alpine

```
leaf-mariadb-1  | s6-rc: info: service ready: starting
leaf-mariadb-1  | s6-rc: info: service ready successfully started
leaf-mariadb-1  | s6-rc: info: service mysql-setup: starting
leaf-mariadb-1  | s6-rc: info: service confd: starting
leaf-mariadb-1  | s6-rc: info: service confd successfully started
leaf-mariadb-1  | Installing MariaDB/MySQL system tables in '/var/lib/mysql' ...
leaf-mariadb-1  | s6-sudoc: fatal: unable to get exit status from server: Operation timed out
leaf-mariadb-1  | s6-rc: warning: unable to start service mysql-setup: command exited 111
leaf-mariadb-1  | /run/s6/basedir/scripts/rc.init: warning: s6-rc failed to properly bring all the services up! Check your logs (in /run/uncaught-logs/current if you have in-container logging) for more information.
leaf-mariadb-1  | prog: fatal: stopping the container.
```

**Comment by nigelgbanks** [comment]
Not to self: Test fails relatively often due to https://www.drupal.org/project/drupal/issues/2833539 occurring when Alpaca tries to write a derivative to Drupal

**Comment by nigelgbanks** [comment]
@jefferya 

> If I set the docker-compose.yml env var S6_CMD_WAIT_FOR_SERVICES_MAXTIME for mariadb 

What value are you setting it to?

**Comment by jefferya** [comment]
> What value are you setting it to?

S6_CMD_WAIT_FOR_SERVICES_MAXTIME: 10min (600000). Same value as in `drupal/Dockerfile`

**Comment by nigelgbanks** [comment]
> > What value are you setting it to?
> 
> S6_CMD_WAIT_FOR_SERVICES_MAXTIME: 10min (600000). Same value as in `drupal/Dockerfile`

I've pushed a change to set that to the default for mariadb as well.

**Comment by nigelgbanks** [comment]
@alxp brought this (https://github.com/Islandora-Devops/isle-dc/issues/298) to my attention during the tech call, so I pushed a fix for that we now have a new image for the [leptonica](https://github.com/Islandora-Devops/isle-leptonica) package, so we can compile it with its optional open jpeg dependency.

**Review by DonRichards — APPROVED** [review]
This built for me without an issue and the derivative look correct.

**Comment by seth-shaw-asu** [comment]
I ran `make setup` without having pre-commit installed (because the README said I didn't need it) but it errors out before completing. After I installed pre-commit it ran completely and notified me that I hadn't installed jq yet. Now that make, pre-commit, and jq are all updated/installed it completes successfully.

**Comment by jefferya** [comment]
I swapped the images into a local review environment (LEAF with mods) and ran a few tests. The site loads and PDF's, video/audio and image ingest looks ok.  I haven't tried Matomo and Fedora in not in LEAF stack. On bit to maybe highlight is the change log Cantaloupe port change -- this caught me for a bit.

**Comment by rosiel** [comment]
Like @seth-shaw-asu, I had to install `pre-commit` and `jq`. Now it seems to work. I ran `make bake` (on mac with homebrew it's `gmake bake` to avoid warnings caused by the older built-in version of make) and then `make certs` and `make up` to launch the test site. I was able to access it and use the site and the IDE (woohoo!) 

Aside from potentially listing Cantaloupe in the "Services available" in `make up`, are we good to proceed?

**Review by rosiel — APPROVED** [review]

**Comment by nigelgbanks** [comment]
> I haven't tried Matomo and Fedora in not in LEAF stack. On bit to maybe highlight is the change log Cantaloupe port change -- this caught me for a bit.

I've updated the README so it's not an explicit requirement

**Comment by nigelgbanks** [comment]
> Aside from potentially listing Cantaloupe in the "Services available" in `make up`, are we good to proceed?

I've pushed a commit to add it to the list of services, good catch.

**Comment by seth-shaw-asu** [comment]
I ran `make certs` and `make up` but Microsoft Edge didn't like the certs and won't let me view the site.

**Comment by nigelgbanks** [comment]
> I ran `make certs` and `make up` but Microsoft Edge didn't like the certs and won't let me view the site.

I'll see what I can do about that `mkcert` doesn't seem to document any additional steps, and I didn't see anything in their issue queue about Edge. So, we may not be able to support Edge with this...

**Comment by nigelgbanks** [comment]
> I ran `make certs` and `make up` but Microsoft Edge didn't like the certs and won't let me view the site.

I've pushed a fix for this. In this particular case, `mkcert` must be run in the host Windows context (i.e. `mkcert.exe` not under WSL2). As it needs to be able to modify the trust store on Windows to affect the browsers running on Windows. Now the makefile will check if it is running under WSL2 and call the appropriate executable.

**Comment by nigelgbanks** [comment]
@seth-shaw-asu if you test again under Windows be sure to delete the existing certificates you generated under `build\certs` to ensure that new ones get created.

**Comment by nigelgbanks** [comment]
A number of people have reviewed and tested, so I feel it's safe to merge. If any issues arise, we can raise new tickets, and I'll sort it out.

**Pull request merged**